STYLEGUIDE.md

```diff
@@ -0,0 +1,716 @@
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use upper case snake case for enums and enum properties */
+enum SomeEnum {
+  FOO_BAR,
+  BAR_FOO
+}
+
+/* Don't use camel case for enums and enum properties */
+enum SomeEnum {
+  fooBar,
+  barFoo
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use pascal case */
+enum SomeEnum {
+  FooBar,
+  BarFoo
+}
+```
+### Use verbs in function names
+Every function should do something and that action should be described with a verb.
+The verb should fit the return type and value.
+An exception to this can be methods, that operate on an instance directly, e. g. `instance.toString()`, `instance.asObject()`.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use functions without verb */
+function toFoo(): Foo {
+  ...
+}
+
+/* Don't use functions with misleading verbs */
+function isFoo(): void {
+  ...
+  throw Error(...);
+}
+
+/* Don't use camel case for enums and enum properties */
+enum SomeEnum {
+  fooBar,
+  barFoo
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use functions with verbs */
+function transformToFoo(): Foo {
+  ...
+}
+
+/* Use functions with verbs that fit the return type */
+function isFoo(): boolean {
+  ...
+}
+
+/* Use functions with verbs that fit the return type */
+function validateFoo(): void {
+  ...
+  throw Error(...);
+}
+
+/* Exception, you can omit verbs in methods, that transform the instance that they are invoked on */
+class FooBar {
+  toString(): string {
+    ...
+  }
+
+  asObject(): Record<string, any> {
+    ...
+  }
+}
+```
+
+## Strings
+
+### Use single quotes ✓
+
+Use single quotes, unless not possible otherwise, e. g. when your string
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use double quotes */
+const foo = "foo";
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use single quotes */
+const foo = 'foo';
+
+/* Use double quotes if the quoted string contains single quotes. */
+const foo = "'foo'";
+```
+
+### String concatenation
+When possible use template literals, this makes the code more readable in most cases.
+Feel free to concatenate strings with `+` in case it makes to code better readable.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't concatenate strings with `+` */
+const foo = 'foo ' + bar + ' bar';
+
+/* Grey area: If template literals become too complex, don't use them. */
+const foo = `foo ${fn(`${bar}bar`)}`;
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Concatenate strings using template literals */
+const foo = `foo ${bar} bar`;
+
+/* Grey area: If template literals become too complex, feel free to concatenate with +. */
+const foo = `foo ${fn(bar +'bar')}`;
+```
+
+## Arrays
+
+### Use `.forEach` rather than `for`
+Use the functional language concepts of JavaScript in favor of the imperative ones, when possible.
+Use the correct function for your use case.
+In rare cases, when performance is of the essence, it makes sense to resort to the imperative `for` instead.
```

Comment:
I actually find it easier to read `for...of` loops and it is easier to use. `forEach` has pretty weird behavior (e.g. returns `undefined`) and can't easily be used with `async/await` (this might be worth a mention in this part of the style guide as well -> prefer `await Promise.all(array.map())`)